### ISSUE TITLE
reading hiera settings

### ISSUE THREAD

**Issue by tuxmea** [example42/hdm#14]
1. a hiera.yaml file is fully optional.
If no file is available, hiera uses internal defaults

2. HDM expects hiera data to be in environmentpath/data
But data can be set using hiera.yaml file:

```
---
version: 5
defaults:
  datadir: modules/hieradata/data
```

Defaults are in hiera ruby gem:
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/hiera/config.rb

```
class Hiera::Config
  class << self
    ##
    # load takes a string or hash as input, strings are treated as filenames
    # hashes are stored as data that would have been in the config file
    #
    # Unless specified it will only use YAML as backend with a
    # hierarchy of 'nodes/%{::trusted.certname}' and 'common', and with a
    # console logger.
    #
    # @return [Hash] representing the configuration.
    def load(source)
      @config = {:backends => ["yaml"],
                 :hierarchy => ["nodes/%{::trusted.certname}", "common"],
                 :merge_behavior => :native }

      if source.is_a?(String)
        if File.exist?(source)
          config = begin
                     yaml_load_file(source)
                   rescue TypeError => detail
                     case detail.message
                     when /no implicit conversion from nil to integer/
                       false
```

**Comment by oneiros** [comment]
I did not realize how complicated the current situation with puppet/hiera is, so I had to take a very deep dive.

The hiera ruby gem is officially "deprecated in favor of Hiera version 5 which is implementation (sic) in Puppet." But there is no version 5. Instead puppet uses the hiera gem and adds a couple of wrappers and monkey-patches.

I am pretty sure that when using puppet, the actual config parsing happens in https://github.com/puppetlabs/puppet/blob/07a73247b693f45393e440b711242391e2cacb46/lib/puppet/pops/lookup/hiera_config.rb.

In there puppet distinguishes between the different hiera config file versions and _uses different defaults for different versions_.

Here are the defaults for v3:

```ruby
DEFAULT_CONFIG_HASH =  {
  KEY_BACKENDS => %w(yaml),
  KEY_HIERARCHY => %w(nodes/%{::trusted.certname} common),
  KEY_MERGE_BEHAVIOR => 'native'
}
```

These mirror the ones seen above.

But for v5 we get a totally different set of defaults:

```ruby
DEFAULT_CONFIG_HASH = {
  KEY_VERSION => 5,
  KEY_DEFAULTS => {
    KEY_DATADIR => 'data',
    KEY_DATA_HASH => 'yaml_data'
  },
  KEY_HIERARCHY => [
    {
      KEY_NAME => 'Common',
      KEY_PATH => 'common.yaml',
    }
  ]
}
```

In case no file is given (or rather an empty configuration is given), it assumes version 3.

So, from my superficial and probably incomplete understanding of puppet's code, I think this is what hdm needs to do:

* If there is no "hiera.yaml", assume version 3 and use version 3 defaults.
* If there is a "hiera.yaml", parse it and merge parsed values with the appropriate defaults depending on the version number.

**Comment by oneiros** [comment]
Looking at how to implement the above I noticed that while there are some references to v3-style configuration in many places there is already a hardcoded expectation to have v5-style data.

@tuxmea Does hdm need to support hiera v3 configuration files?

**Comment by tuxmea** [comment]
@oneiros : We only want hiera v5 support. No hiera v3 support needed in HDM.

When there is no hiera.yaml file, HDM should not do anything and only show an error, that hiera.yaml file is needed.
If there is a hiera.yaml file, check that it has `version: 5` inside. Otherwise show an error.

**Comment by tuxmea** [comment]
implemented in #42
